src/Language/PureScript/Parser/Declarations.hs

```diff
@@ -448,8 +448,8 @@ parseValueAtom = withSourceSpan PositionedValue $ P.choice
                  , P.try parseVar
                  , parseCase
                  , parseIfThenElse
-                 , parseDo
-                 , parseAdo
+                 , P.try parseDo
+                 , P.try parseAdo
```

Comment:
I don't think we want to add backtracking here. This will backtrack on _any_ failure inside a do-block, which will result in poor errors. We only want to apply backtracking to the parsing of the `do` keyword, which means we need to move this into `parseDo` and `parseAdo`.

Reply:
`<|>` will only fallback to the next branch if you have wrapped in `try` or you have not consumed any tokens. `parseQualified` likely consumes the qualification, so when it does not get a `do`, it will _not_ fallback to the next alternative unless you explicitly wrap it in `try`.